pkgs/pkgs-lib/formats/hocon/src/src/main.rs

```diff
@@ -0,0 +1,234 @@
+                let required = o
+                    .get("required")
+                    .unwrap_or(&Value::Bool(false))
+                    .as_bool()
+                    .unwrap_or_else(|| panic!("Substition value is not a string: {:?}", o));
+
+                debug_assert!(!value.contains("}"));
+
+                HOCONValue::Substitution(value.to_string(), required)
+            }
+            "append" => {
+                let value = o.get("value").expect("Missing value for append");
+
+                HOCONValue::Append(Box::new(json_to_hocon(value)))
+            }
+            _ => panic!(
+                "\
+          Attribute set contained special element '_type',\
+          but its value is not recognized:\n{}",
+                r#type
+            ),
+        })
+}
+
+fn json_to_hocon(v: &Value) -> HOCONValue {
+    match v {
+        Value::Null => HOCONValue::Null,
+        Value::Bool(b) => HOCONValue::Bool(*b),
+        Value::Number(n) => {
+            if n.is_i64() {
+                HOCONValue::Int(n.as_i64().unwrap())
+            } else if n.is_f64() {
+                HOCONValue::Float(n.as_f64().unwrap())
+            } else {
+                panic!(
+                    "{} is not i64 or f64, cannot be represented as number in HOCON",
+                    n
+                );
+            }
+        }
+        Value::String(s) => HOCONValue::String(s.to_string()),
+        Value::Array(a) => {
+            let items = a
+                .iter()
+                .map(json_to_hocon)
+                .collect::<Vec<HOCONValue>>();
+            HOCONValue::List(items)
+        }
+        Value::Object(o) => {
+            if let Some(result) = parse_special_types(o) {
+                return result;
+            }
+
+            let mut items = o
+                .iter()
+                .filter(|(key, _)| key.as_str() != "_includes")
+                .map(|(key, value)| (key.clone(), json_to_hocon(value)))
+                .collect::<Vec<(String, HOCONValue)>>();
+
+            items.sort_by(|(a, _), (b, _)| a.partial_cmp(b).unwrap());
+
+            let includes = o
+                .get("_includes")
+                .map(|x| {
+                    x.as_array()
+                        .expect("_includes is not an array")
+                        .iter()
+                        .map(|x| x.as_object().expect(format!("Include is not an object: {}", x).as_str()))
+                        .map(parse_include)
+                        .collect::<Vec<HOCONInclude>>()
+                })
+                .unwrap_or(vec![]);
+
+            HOCONValue::Object(includes, items)
+        }
+    }
+}
+
+impl ToString for HOCONValue {
+    fn to_string(&self) -> String {
+        match self {
+            HOCONValue::Null => "null".to_string(),
+            HOCONValue::Bool(b) => b.to_string(),
+            HOCONValue::Int(n) => n.to_string(),
+            HOCONValue::Float(n) => format!("{:?}", n),
```

Comment:
Are `Debug` implementations guaranteed to be stable in Rust? Otherwise can this be `{}` instead?

Reply:
The debug implementations for structs in `std` are unfortunately marked as unstable. The reason I'm using the debug for float specifically, is because the normal `Display` implementation for some reason drops `.0` for floats that are whole numbers. When parsed by HOCON, this would then change its type.

I was playing with the idea of copying the implementation to pin it, but it seems to depend on a tree of abstractions that is not exposed, so it feels like it costs more than the reward. Even though it's marked as unstable, I think they are unlikely to change any time soon.

Reply:
According to https://github.com/lightbend/config/blob/main/HOCON.md, it seems to use the same numbers as JSON, including the fact that there's no distinction between floats and ints:

> - values have possible types: string, number, object, array, boolean, null
> - allowed number formats matches JSON; as in JSON, some possible floating-point values are not represented, such as NaN

So it seems like it should be fine to ditch `HOCONValue::{Int,Float}` in favor of `HOCONValue::Number` and use `serde_json` to serialise its value.